cpp/src/arrow/builder.cc

```diff
@@ -872,8 +894,124 @@ Status BooleanBuilder::Append(const std::vector<bool>& values) {
 // ----------------------------------------------------------------------
 // DictionaryBuilder
 
+using internal::DictionaryScalar;
 using internal::WrappedBinary;
 
+namespace {
+
+// A helper class to manage a hash table embedded in a typed Builder.
+template <typename T, typename Enable = void>
+struct DictionaryHashHelper {};
+
+// DictionaryHashHelper implementation for primitive types
+template <typename T>
+struct DictionaryHashHelper<T,
+                            typename std::enable_if<TypeTraits<T>::is_primitive>::type> {
+  using Builder = typename TypeTraits<T>::BuilderType;
+  using Scalar = typename DictionaryScalar<T>::type;
+
+  // Get the dictionary value at the given builder index
+  static Scalar GetDictionaryValue(const Builder& builder, int64_t index) {
```

Comment:
Do you know if `inline` is redundant here? I think so but just confirming my understanding of the C++ spec cc @cpcloud @joshuastorck

Reply:
@wesm do you mean `static`? if so, then yes `static` is redundant inside of an anonymous namespace. Both anonymous namespaces and `static` are two different ways of declaring [internal linkage](http://en.cppreference.com/w/cpp/language/storage_duration#Linkage).

Reply:
Sorry I missed that this `static` is on the method, in which case it is of course meaningful beyond anonymous namespaces.

Reply:
No I mean `inline`. Because this function is in a templated struct that is used later, is `static` the same as `static inline`

Reply:
I think any method defined inside the `struct` or `class` definition is considered `inline`?

Reply:
When you declare complete member functions in a class declarations, they are implicitly inline (so writing `inline` is redundant), I believe that is the same here but just want to confirm

Reply:
From http://en.cppreference.com/w/cpp/language/inline:

> A function defined entirely inside a class/struct/union definition, whether it's a member function or a non-member friend function, is implicitly an inline function.

Reply:
:+1: great, thank you =)